### PR TITLE
ISPN-9687 Simple remote query

### PR DESCRIPTION
**Pull request by karesti**
https://issues.jboss.org/browse/ISPN-9687

### CODE REVIEW

**Comment by karesti** [comment]
@anistor I'm getting a message WARN: Encountered and ignored and unknown annotation "Doc" on tutorial.InstaPost
What should we put ?

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+         "infinispan");
+
+   private static final String CACHE_NAME = "instaposts";
+
+   public static void main(String[] args) throws Exception {
+      // Create a configuration for a locally-running server
+      ConfigurationBuilder builder = new ConfigurationBuilder();
+      builder.addServer().host("127.0.0.1")
+            .port(ConfigurationProperties.DEFAULT_HOTROD_PORT)
+            .clientIntelligence(ClientIntelligence.BASIC) // This line is needed to connect with docker4mac
```
Comment (galderz) [inline remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java]: Hmmm, if the host is `127.0.0.1` do you need client intelligence basic? I'd remove basic intelligence from this tutorial.

Comment (karesti) [reply remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java]: ok

**Comment by karesti** [comment]
done @galderz

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InstaPost.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * This class is annotated with the infinispan Protostream support annotations. With this method, you don't need to
+ * define a protobuf file and a marshaller for the object. See this tutorial for additional explanations:
+ * <p>
+ * https://blog.infinispan.org/2018/06/making-java-objects-queryable-by.html
+ */
+@ProtoDoc("@Indexed")
+public final class InstaPost {
+
+   @ProtoDoc("@Field")
+   @ProtoField(number = 1, required = true)
```
Comment (anistor) [inline remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InstaPost.java]: Required fields are going away in the future. They are still supported ATM of course but it would be ideal not to use them in tutorials to make them future-proof.

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InstaPost.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+   public InstaPost(String id, String user, String hashtag) {
+      this.id = id;
+      this.user = user;
+      this.hashtag = hashtag;
+   }
+
+   @Override
+   public String toString() {
+      return "Tag{" +
```
Comment (anistor) [inline remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InstaPost.java]: Should be "InstaPost" actually.

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+      // Create the continuous query
+      ContinuousQuery<String, InstaPost> continuousQuery = Search.getContinuousQuery(instaPostsCache);
+
+      // Create the continuous query listener.
+      ContinuousQueryListener<String, InstaPost> listener =
+            new ContinuousQueryListener<String, InstaPost>() {
+               // This method will be executed every time new items that correspond with the query arrive
+               @Override
+               public void resultJoining(String id, InstaPost post) {
```
Comment (anistor) [inline remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java]: Id call this ```key```, not ```id```.

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+      // Add 1000 random posts
+      for (int i = 0; i < 1000; i++) {
+         // Add a post
+         addRandomPost(instaPostsCache);
+
+         // Await a little to see results
+         Thread.sleep(10);
+      }
+
+      // Clear the cache
```
Comment (anistor) [inline remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java]: This actually removes the cache instead of just clearing it  :)

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+         "infinispan");
+
+   private static final String CACHE_NAME = "instaposts";
+
+   public static void main(String[] args) throws Exception {
+      // Create a configuration for a locally-running server
+      ConfigurationBuilder builder = new ConfigurationBuilder();
+      builder.addServer().host("127.0.0.1")
+            .port(ConfigurationProperties.DEFAULT_HOTROD_PORT)
+            .marshaller(ProtoStreamMarshaller.class); // You need to specify the marshaller
```
Comment (anistor) [inline remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java]: ```// You need to specify the marshaller```  -> ```// You need to specify the marshaller for remote query```

File: remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+ * @author Katia Aresti, karesti@redhat.com
+ */
+public class InfinispanRemoteQuery {
+
+   public static void main(String[] args) throws Exception {
+      // Create a configuration for a locally-running server
+      ConfigurationBuilder builder = new ConfigurationBuilder();
+      builder.addServer().host("127.0.0.1")
+            .port(ConfigurationProperties.DEFAULT_HOTROD_PORT)
+            .marshaller(ProtoStreamMarshaller.class); // You need to specify the marshaller
```
Comment (anistor) [inline remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java]: ```// You need to specify the marshaller``` -> ```// You need to specify the marshaller for remote query```

File: remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.RemoteCacheManager;
+import org.infinispan.client.hotrod.Search;
+import org.infinispan.client.hotrod.configuration.ClientIntelligence;
```
Comment (anistor) [inline remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java]: Not used.

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Random;
+import java.util.UUID;
+
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.RemoteCacheManager;
+import org.infinispan.client.hotrod.Search;
+import org.infinispan.client.hotrod.configuration.ClientIntelligence;
```
Comment (anistor) [inline remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java]: Not used.

File: remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+            .marshaller(ProtoStreamMarshaller.class); // You need to specify the marshaller
+
+      // Connect to the server
+      RemoteCacheManager client = new RemoteCacheManager(builder.build());
+
+      // Get the persons cache, create it if needed with the default configuration
+      RemoteCache<String, Person> personsCache = client.administration().getOrCreateCache("people", "default");
+
+      // Create the persons dataset to be stored in the cache
+      Map<String, Person> persons = Stream.of(new Object[][]{
```
Comment (anistor) [inline remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java]: Stream abuse. I'll call the Java police! :))

File: remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+
+      // Print the results
+      System.out.println(rossignols);
+
+      // Stop the client and release all resources
+      client.stop();
+   }
+
+   private static void addPersonSchema(RemoteCacheManager cacheManager) throws IOException {
+      // Get the serialization context
```
Comment (anistor) [inline remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java]: // Get the serialization context of the client

File: remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+      SerializationContext ctx = ProtoStreamMarshaller.getSerializationContext(cacheManager);
+
+      // Use ProtoSchemaBuilder to define
```
Comment (anistor) [inline remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java]: // Use ProtoSchemaBuilder to define a Protobuf schema on the client

File: remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+            .fileName(fileName)
+            .addClass(Person.class)
+            .packageName("tutorial")
+            .build(ctx);
+
+      // Retrieve metadata cache
+      RemoteCache<String, String> metadataCache =
+            cacheManager.getCache(PROTOBUF_METADATA_CACHE_NAME);
+
+      // Store the configuration in the metadata cache
```
Comment (anistor) [inline remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java]: ```configuration``` is not a suitable term. I'd rephrase this as "// Define the new schema on the server also"

Comment (anistor) [reply remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java]: These two comments apply to the other tutorial as well.

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InstaPost.java (hunk 2 above)
Comment (karesti) [reply remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InstaPost.java]: how should I do it then ?

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java (hunk 5 above)
Comment (karesti) [reply remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java]: changing this the comment

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InstaPost.java (hunk 3 above)
Comment (karesti) [reply remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InstaPost.java]: ok

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java (hunk 4 above)
Comment (karesti) [reply remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java]: ok

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java (hunk 6 above)
Comment (karesti) [reply remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java]: ok

File: remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java (hunk 7 above)
Comment (karesti) [reply remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java]: ok

File: remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java (hunk 8 above)
Comment (karesti) [reply remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java]: ok

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java (hunk 9 above)
Comment (karesti) [reply remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java]: ok

File: remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java (hunk 10 above)
Comment (karesti) [reply remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java]: okkkkk haha 👮

File: remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java (hunk 12 above)
Comment (karesti) [reply remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java]: ok

File: remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java (hunk 13 above)
Comment (karesti) [reply remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java]: ok

File: remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java (hunk 11 above)
Comment (karesti) [reply remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java]: ok

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InstaPost.java (hunk 2 above)
Comment (karesti) [reply remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InstaPost.java]: if I remove required I get this error
Exception in thread "main" ```org.infinispan.protostream.annotations.ProtoSchemaBuilderException: Primitive field 'bornYear' of class org.infinispan.tutorial.simple.remote.query.Person should be marked required or should have a default value.```

I have to provide a default value instead?

**Comment by karesti** [comment]
@anistor I reported the changes

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+            new ContinuousQueryListener<String, InstaPost>() {
+               // This method will be executed every time new items that correspond with the query arrive
+               @Override
+               public void resultJoining(String key, InstaPost post) {
+                  System.out.println(String.format("@%s has posted again! Hashtag: #%s", post.user, post.hashtag));
+                  queryPosts.add(post);
+               }
+            };
+      // Clear all the listeners
+      continuousQuery.removeAllListeners();
```
Comment (anistor) [inline remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java]: Why do we need this ?

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+            .port(ConfigurationProperties.DEFAULT_HOTROD_PORT)
+            .marshaller(ProtoStreamMarshaller.class); // You need to specify the marshaller for remote query
+
+      // Connect to the server
+      RemoteCacheManager client = new RemoteCacheManager(builder.build());
+
+      // Get the cache, create it if needed with the default configuration
+      RemoteCache<String, InstaPost> instaPostsCache = client.administration().getOrCreateCache(CACHE_NAME, "default");
+
+      // Create and add the marshalling configuration for Person pojo class. Note Person is an annotated POJO
```
Comment (anistor) [inline remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java]: ```// Create and add the Protobuf schema for InstaPost class. Note InstaPost is an annotated POJO```

File: remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+      RemoteCache<String, Person> personsCache = client.administration().getOrCreateCache("people", "default");
+
+      // Create the persons dataset to be stored in the cache
+      Map<String, Person> persons = new HashMap<>();
+      persons.put("1", new Person("Oihana", "Rossignol", 2016, "Paris"));
+      persons.put("2", new Person("Elaia", "Rossignol", 2018, "Paris"));
+      persons.put("3", new Person("Yago", "Steiner", 2013, "Saint-Mandé"));
+      persons.put("4", new Person("Alberto", "Steiner", 2016, "Paris"));
+
+      // Create and add the marshalling configuration for Person pojo class. Note Person is an annotated POJO
```
Comment (anistor) [inline remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java]: marshalling configuration? nope, protobuf schema: ```// Create and add the Protobuf schema for Person class. Note Person is an annotated POJO```

Comment (karesti) [reply remote-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteQuery.java]: oki

File: remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java (hunk 15 above)
Comment (karesti) [reply remote-continuous-query/src/main/java/org/infinispan/tutorial/simple/remote/query/InfinispanRemoteContinuousQuery.java]: ok

**Comment by karesti** [comment]
@anistor reported the last comments

**Comment by galderz** [comment]
All good for me, @anistor happy with fixes?

**Comment by anistor** [comment]
Almost, @galderz . I think we do not need to define default values for those nullable fields, but @karesti noticed some issues with that. I'll have to check that myself. And I think we must squash these individual commits and keep just 2, one for query, one for continuous query.

**Comment by anistor** [comment]
Just checked and those default values are not need. I remove them, I squashed the commits and integrated this. Thanks @karesti and @galderz !
